### ISSUE TITLE
Slowness when pairing smaller width with long strings

### ISSUE THREAD

**Issue by filip-gomore** [kkinnear/zprint#290]
I was hoping you would be able to help us with a performance issue when we use `zprint` to format our Clojure code. In some files we are experiencing slowness (takes a few seconds to format the file) and I think it's connected to the string lengths because they can't get broken up. It looks hard for `zprint` to rule out branches and it seems to try a lot of options until it finds a solution.

Here is a prototypical example of how the code looks in those files (I just ran an obfuscation on a real function):

```clojure
(defn lysnxrdilceshiq
  [{:keys [bpwwdtdrbcbeqhtzpm lbbfwa scrx as]}
   {:keys [suufsw raerqzadu] :as vjzoqll}]
  (let [zcxjsmvvgtjdjhgb (z (:rh gjnz) (:rehsrqsjammnytt fakm))
        pbaaxrgfwlf      (:rgjnpzgkigr opkf)]
    {:yniykpynbytpcl
     (isvmgl
      aezz
      [{:wmujdsnfal "ludrasz"
        :oqxmk (xi "dkualogsvwfdlodqbhpkymsosayibtdgggagmbwukjj")
        :uwdwj
        (zsnkrm
         xxjk
         [(when (vues (:yw xulg))
            (whvkjcahz
             "ctrhhbi"
             (rp "gdndmfcqbgeclpmalnjkgzdaceuoufjemuszxnqhqjusxygeblrkimafbo"
                 {:bs techhugyrzuyjqfh})
             (grtpccxmsjnm (:szpdnoueqmzoa twdk) vflrimegn hoeowpljulp)
             {:clbvjylm
              (qt
               "hwdxyisaezgwplnfljfnvyildhsjbggkbplkwjpszemjvyhawdmrfumitlowu"
               {:edofyanir (:qo udbs)
                :qvwepdygi (r (:jsyqczfxlxrzfcg ciba) (:zy ncbz))})}))])}])}))
```
and the `.zprint.edn`
```clojure
{:style       [:community :justified :map-nl :pair-nl :binding-nl :respect-bl]
 :width       80
 :fn-force-nl #{:arg1-body :binding :fn}
 :map         {:comma?    false
               :sort?     true
               :force-nl? false
               :justify   {:max-variance 40}}
 :pair        {:force-nl? true :justify {:max-variance 40}}
 :binding     {:justify {:max-variance 40} :justify? true}
 :vector      {:wrap? false}
 :fn-map      {;; clojure.core and special forms
               "def"           :arg2
               "do"            :flow-body
               "try"           :flow-body
               "finally"       :flow-body}}
```
It takes around 1s to format the above. It's a lot faster when I change the `:width` to `120`.

Thanks!

**Comment by filip-gomore** [comment]
Thank you for taking the time to look into this and explain it so thoroughly! I can verify that setting `max-variance` to 0 improves performance for the whole file dramatically. We'll play around with this next week and I'm sure we'll find a solution that works for us.

**Comment by filip-gomore** [comment]
I'll close as we managed to improve the config for our use case (with a width set to `100` now and a few exceptions). I appreciate this tool a lot, thank you for building it!

**Comment by kkinnear** [comment]
I have released `1.2.6`.  I found several places where the justification code was missing optimizations that were available, and fixed those.  Where above I saw a 7 times slowdown when doing justification using `:max-variance 20`, now I see a slowdown of <2.  Which is a lot better, but still slower.

I also added a new configuration option: `:max-depth 100` to all of the `:justify` key values.  Part of what was an issue in your example was that there were things to justify that were pretty deeply nested.  If you run into this again, you could try setting `:max-depth 4` or maybe 5 or something and see if it helps. I think it will.  It will only justify down to `:max-depth` and then it won't even try.  But I hope you don't need that.

Thanks for the appreciation.  It means a lot!
